File: chains/cpp/dexvm/src/gate.cpp
// Copyright (C) 2026, Lux Industries Inc. All rights reserved.
// SPDX-License-Identifier: BSD-3-Clause-Eco

#include "lux/dexvm/gate.hpp"

#include "lux/dexvm/forbidden.hpp"

#include <optional>

namespace lux::dexvm {

std::string_view to_string(NetworkClass c) {
    switch (c) {
        case NetworkClass::Mainnet: return "mainnet";
        case NetworkClass::Testnet: return "testnet";
        default:                    return "dev";
    }
}

bool value_bearing(NetworkClass c) {
    return c == NetworkClass::Mainnet || c == NetworkClass::Testnet;
}

NetworkClass network_class_for(std::uint32_t network_id) {
    switch (network_id) {
        case 1: return NetworkClass::Mainnet;
        case 2: return NetworkClass::Testnet;
        default: return NetworkClass::Dev;
    }
}

std::vector<AssetKind> DexAssetPolicy::kinds() const {
    if (allowed_asset_kinds.empty())
        return {AssetKind::EVMNative, AssetKind::ERC20, AssetKind::UTXO};
    return allowed_asset_kinds;
}

bool DexAssetPolicy::any_synthetic_flag() const {
    return allow_synthetic_assets || allow_synthetic_markets || allow_mock_liquidity;
}

DexAssetPolicy default_dex_asset_policy() {
    DexAssetPolicy p;
    p.allowed_asset_kinds = {AssetKind::EVMNative, AssetKind::ERC20, AssetKind::UTXO};
    return p;
}

Result<void> refuse_under_synthetic_config(
    NetworkClass network_class, const DexAssetPolicy& policy, const Registry& reg,
    const std::function<std::string(const Id&)>& chain_label_for) {
    auto b = [](bool v) { return std::string(v ? "true" : "false"); };

    // (1) the allowed kinds must be a subset of the three real kinds.
    for (AssetKind k : policy.kinds()) {
        if (!valid(k))
            return fail(Err::BadAllowedKind, "\"" + std::string(to_string(k)) + "\"");
    }

    // (2) no synthetic flags on a value-bearing network.
    if (value_bearing(network_class) && policy.any_synthetic_flag())
        return fail_note(Err::SyntheticOnValueNet,
                         "network=" + std::string(to_string(network_class)) + " synthAssets=" +
                             b(policy.allow_synthetic_assets) + " synthMarkets=" +
                             b(policy.allow_synthetic_markets) + " mockLiq=" +
                             b(policy.allow_mock_liquidity));

    // (3) deny-scan every registered asset. Run before the market scan so a
    // tainted asset is reported at its source rather than at a market over it.
    std::optional<Error> scan_err;
    reg.each([&](const Id& id, const Asset& a) {
        if (scan_err) return;
        const std::string label = chain_label_for ? chain_label_for(a.chain_id) : std::string();
        if (auto r = assert_no_forbidden_asset_refs(a, label); !r)
            scan_err = r.error().wrap("registered asset " + cb58(id));
    });
    if (scan_err) return std::unexpected(*scan_err);

    // (4) every ENABLED market must resolve both sides to a registered, enabled
    // asset on the matching network. A market over a synthetic asset has no
    // AssetID to resolve, so this catches it structurally.
    reg.each_market([&](const Id& id, const Market& m) {
        if (scan_err || !m.enabled) return;
        auto base = reg.must_resolve_enabled(m.base_asset_id);
        if (!base) {
            scan_err = Error{Err::EnabledMarketUnknownAsset,
                             std::string(text_of(Err::EnabledMarketUnknownAsset)) + ": market " +
                                 cb58(id) + " base: " + base.error().text};
            return;
        }
        auto quote = reg.must_resolve_enabled(m.quote_asset_id);
        if (!quote) {
            scan_err = Error{Err::EnabledMarketUnknownAsset,
                             std::string(text_of(Err::EnabledMarketUnknownAsset)) + ": market " +
                                 cb58(id) + " quote: " + quote.error().text};
            return;
        }
        if (m.network_id != base->network_id || m.network_id != quote->network_id)
            scan_err = Error{Err::EnabledMarketUnknownAsset,
                             std::string(text_of(Err::EnabledMarketUnknownAsset)) + ": market " +
                                 cb58(id) + " network mismatch (market=" +
                                 std::to_string(m.network_id) + " base=" +
                                 std::to_string(base->network_id) + " quote=" +
                                 std::to_string(quote->network_id) + ")"};
    });
    if (scan_err) return std::unexpected(*scan_err);
    return {};
}

}  // namespace lux::dexvm

File: chains/rust/dexvm/src/gate.rs
// SPDX-License-Identifier: BSD-3-Clause-Eco
// Copyright (C) 2019-2026, Lux Industries Inc. All rights reserved.

//! The single fail-closed startup gate.
//!
//! It runs once, before the chain accepts work, over the already-populated
//! registry. It returns success ONLY when the registry is fully real and the
//! policy is locked down for the network. Anything ambiguous fails closed.

use crate::asset::AssetKind;
use crate::error::{Code, Error, Result};
use crate::forbidden::assert_no_forbidden_asset_refs;
use crate::ids::{self, Id};
use crate::registry::Registry;
use serde::{Deserialize, Serialize};

/// Which networks forbid a synthetic flag outright.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum NetworkClass {
    /// devnet / localnet — synthetic flags MAY be set.
    Dev,
    /// testnet — synthetic flags FORBIDDEN.
    Testnet,
    /// mainnet — synthetic flags FORBIDDEN.
    Mainnet,
}

impl NetworkClass {
    /// Whether this class forbids any synthetic flag. Mainnet and testnet both
    /// carry value for the purpose of this gate.
    fn value_bearing(self) -> bool {
        matches!(self, NetworkClass::Mainnet | NetworkClass::Testnet)
    }
}

impl std::fmt::Display for NetworkClass {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_str(match self {
            NetworkClass::Mainnet => "mainnet",
            NetworkClass::Testnet => "testnet",
            NetworkClass::Dev => "dev",
        })
    }
}

/// The class of a Lux networkID, using the convention-fixed ids (1 mainnet,
/// 2 testnet, 3 local, 1337 localnet). Every other id — 3, 1337 and every
/// sovereign L1 id — is dev-class for the purpose of synthetic-flag permission.
/// Value on those is still guarded by the consensus-mode guard; this class only
/// governs the synthetic flags.
pub fn network_class_for(network_id: u32) -> NetworkClass {
    match network_id {
        1 => NetworkClass::Mainnet,
        2 => NetworkClass::Testnet,
        _ => NetworkClass::Dev,
    }
}

/// The backend-enforced, fail-closed configuration for the DEX's
/// real-assets-only posture.
///
/// Every field's SAFE value is its default, so a default-initialised policy is
/// the locked-down policy. These are not front-end toggles: they live in the VM
/// config and are read once at initialisation. A front end cannot relax them.
#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize)]
pub struct DexAssetPolicy {
    /// On a value-bearing network any true value fails startup. On a dev network
    /// a true value is permitted — developer opt-in — but is still subject to
    /// the forbidden-reference scan: an off-network universe is never allowed.
    #[serde(rename = "dexAllowSyntheticAssets", default)]
    pub allow_synthetic_assets: bool,
    #[serde(rename = "dexAllowSyntheticMarkets", default)]
    pub allow_synthetic_markets: bool,
    #[serde(rename = "dexAllowMockLiquidity", default)]
    pub allow_mock_liquidity: bool,
    /// The active `dexAllowedAssetKinds` set. Empty is the canonical default of
    /// all three; an explicit set may only ever be a SUBSET of them.
    #[serde(rename = "dexAllowedAssetKinds", default)]
    pub allowed_asset_kinds: Vec<AssetKind>,
}

impl DexAssetPolicy {
    /// The canonical locked-down policy: no synthetic anything, all three real
    /// kinds allowed.
    pub fn default_policy() -> DexAssetPolicy {
        DexAssetPolicy {
            allow_synthetic_assets: false,
            allow_synthetic_markets: false,
            allow_mock_liquidity: false,
            allowed_asset_kinds: vec![AssetKind::EvmNative, AssetKind::Erc20, AssetKind::Utxo],
        }
    }

    /// The effective allowed-kind set: the canonical three when unset, else
    /// exactly the configured subset.
    pub fn allowed_kinds_or_default(&self) -> Vec<AssetKind> {
        if self.allowed_asset_kinds.is_empty() {
            vec![AssetKind::EvmNative, AssetKind::Erc20, AssetKind::Utxo]
        } else {
            self.allowed_asset_kinds.clone()
        }
    }

    /// Whether any synthetic or mock flag is set. The node uses it to
    /// MACHINE-DERIVE the labeled-launch "real-assets-only" assertion rather
    /// than trusting an operator-supplied claim.
    pub fn any_synthetic_flag(&self) -> bool {
        self.allow_synthetic_assets || self.allow_synthetic_markets || self.allow_mock_liquidity
    }
}

/// Refuse startup under a synthetic configuration.
///
/// This is called once at initialisation, BEFORE the chain accepts work, with
/// the already-populated registry and the VM's network class and policy. It
/// refuses if ANY of the following hold:
///
/// 1. `dexAllowedAssetKinds` contains a non-real kind — it may only ever be a
///    subset of the three.
/// 2. The network is value-bearing AND any synthetic flag is true.
/// 3. Any registered asset carries a forbidden reference: an off-network
///    white-label universe chain, mock liquidity, or an ASCII-ticker id. The
///    positive reality check already happened at registration; this is the
///    residual deny-scan, which reality alone would not catch.
/// 4. Any ENABLED market references an asset that is not registered, is
///    disabled, or spans a network mismatch.
///
/// `chain_label_for` maps a source chain id to its human label so the
/// off-network scan can run; a chain with no label simply cannot be a known
/// white-label universe, and its asset already passed the reality gate.
pub fn refuse_under_synthetic_config(
    class: NetworkClass,
    policy: &DexAssetPolicy,
    reg: &Registry,
    chain_label_for: Option<&dyn Fn(Id) -> String>,
) -> Result<()> {
    // (1) allowed-kinds must be a subset of the three real kinds.
    for k in policy.allowed_kinds_or_default() {
        if !k.valid() {
            return Err(Error::detail(
                Code::BadAllowedKind,
                format!("{:?}", k.as_str()),
            ));
        }
    }

    // (2) no synthetic flags on a value-bearing network.
    if class.value_bearing() && policy.any_synthetic_flag() {
        return Err(Error::note(
            Code::SyntheticOnValueNet,
            format!(
                "network={class} synthAssets={} synthMarkets={} mockLiq={}",
                policy.allow_synthetic_assets,
                policy.allow_synthetic_markets,
                policy.allow_mock_liquidity
            ),
        ));
    }

    // (3) deny-scan every registered asset. Run before the market scan so a
    // tainted asset is reported at its source.
    for (id, a) in reg.assets() {
        let label = chain_label_for.map(|f| f(a.chain_id)).unwrap_or_default();
        assert_no_forbidden_asset_refs(&a.symbol, &a.name, &label)
            .map_err(|e| e.wrap(format!("registered asset {}", ids::cb58(&id))))?;
    }

    // (4) every ENABLED market must resolve both sides to a registered, enabled
    // asset on the matching network. A market over a synthetic asset has no
    // AssetID to resolve, so this catches it structurally.
    for (id, m) in reg.markets() {
        if !m.enabled {
            continue;
        }
        let base = reg.must_resolve_enabled(&m.base_asset_id).map_err(|e| {
            Error::detail(
                Code::EnabledMarketUnknownAsset,
                format!("market {} base: {}", ids::cb58(&id), e),
            )
        })?;
        let quote = reg.must_resolve_enabled(&m.quote_asset_id).map_err(|e| {
            Error::detail(
                Code::EnabledMarketUnknownAsset,
                format!("market {} quote: {}", ids::cb58(&id), e),
            )
        })?;
        if m.network_id != base.network_id || m.network_id != quote.network_id {
            return Err(Error::detail(
                Code::EnabledMarketUnknownAsset,
                format!(
                    "market {} network mismatch (market={} base={} quote={})",
                    ids::cb58(&id),
                    m.network_id,
                    base.network_id,
                    quote.network_id
                ),
            ));
        }
    }
    Ok(())
}
